Adopt UIResponder API methods in lieu of handleKeyUIEvent to respond to press events

aprotyas opened the pull request:
#### a5fcaa3c441c0f15061f29b8bc50bcd8198eb28f
<pre>
Adopt UIResponder API methods in lieu of handleKeyUIEvent to respond to press events
<a href="https://bugs.webkit.org/show_bug.cgi?id=260217">https://bugs.webkit.org/show_bug.cgi?id=260217</a>
rdar://100726289

Reviewed by Wenson Hsieh.

The handleKeyUIEvent: and _handleKeyUIEvent: SPI are both deprecated.
Instead, we choose to adopt the pressesBegan:, pressesChanged:,
pressesEnded:, and pressesCancelled: API in this commit. These methods
are forwarded events for handling in the same fashion as the SPI to be
removed.

* LayoutTests/fast/repaint/placeholder-after-caps-lock-hidden.html:

After the changes introduced in this patch, this test was dumping a lot
more painted rects. This was because previous key actions (`a`/`\b`)
were bleeding into the section of the test where we wanted to track
repaints (i.e. after the capslock indicator has been hidden).

We fix this by enforcing correct event ordering through appropriate async
usage of the UIHelper functions (i.e. awaiting `keyDown()` calls and the
resulting `keyUp` events).

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

1. Remove the deprecated SPI declarations.
2. Declare the UIPressInfo interface and some associated methods to
imbibe event information into a UIPress object.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _tryToHandleKeyEventInCustomContentView:]):

Helper function to determine whether an event needs forwarding based on
whether or not a custom content view handles it. Helps deduplicate code.

(-[WKWebView pressesBegan:withEvent:]):
(-[WKWebView pressesChanged:withEvent:]):
(-[WKWebView pressesEnded:withEvent:]):
(-[WKWebView pressesCancelled:withEvent:]):
(-[WKWebView _handleKeyUIEvent:]): Deleted.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _tryToHandlePressesEvent:]):

Helper function to determine whether an event needs forwarding based on
whether or not an input peripheral handles it. Helps deduplicate code.

(-[WKContentView pressesBegan:withEvent:]):
(-[WKContentView pressesEnded:withEvent:]):
(-[WKContentView pressesChanged:withEvent:]):
(-[WKContentView pressesCancelled:withEvent:]):
(-[WKContentView _handleKeyUIEvent:]): Deleted.

(WTR::UIScriptControllerIOS::toggleCapsLock):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::createUIPhysicalKeyboardEvent):
(WTR::UIScriptControllerIOS::toggleCapsLock):

The `toggleCapsLock` implementation required two broad changes to
maintain the existing toggle behavior:

1. Simply calling `pressesBegan:` on the `UIApplication` singleton does
not suffice in getting our synthetic keyboard event sent to the content
view&apos;s overridden `pressesBegan:` definition. To work around this, we
manually create a `UIPress` object with knowledge of our synthetic
keyboard event, and then send the press/event combination directly to
our content view&apos;s `pressesBegan:` overload. This maintains the existing
toggle behavior.

2. In inspecting the press/event combinations received when toggling the
capslock key with a hardware keyboard, it was noticed that on iOS we send
both a keyDown and a keyUp event with the appropriate modifier flag
information whenever we want to toggle capslock, with the former being
sent to `pressesBegan:` and the latter being sent to `pressesEnded:`.
As such, we generate a pair of keyboard events and call the appropriate
`presses[Began/Ended]:` method directly on our content view to reflect
this behavior.

Canonical link: <a href="https://commits.webkit.org/267119@main">https://commits.webkit.org/267119@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a9f0c0e3299c9577dad8b8bdba654f208a87d777

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16442 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17512 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14780 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15935 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16163 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15940 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16409 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/13408 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18265 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13659 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14219 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21123 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14384 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17641 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14979 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14226 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/14059 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3756 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18595 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14801 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -1356,9 +1359,19 @@ static UIDeviceOrientation toUIDeviceOrientation(DeviceOrientation orientation)
+        auto *keyboardEventUp = createUIPhysicalKeyboardEvent(@"capsLock", [NSString string], capsLockOn ? UIKeyModifierAlphaShift : 0, kUIKeyboardInputModifierFlagsChanged, IsKeyDown::No);
+        return std::make_pair(keyboardEventDown, keyboardEventUp);
+    }();
+    auto *press = [&keyboardEvent = keyboardEventDown] {
```
This whole dance because it's illegal to (explicitly or implicitly) capture a structured binding in C++
https://burnicki.pl/en/2021/04/19/capture-structured-bindings.html

pxlcoder replied:
Why do we need lambdas here? (as opposed to just using local variables)

aprotyas replied:
No particular reason really, just the way I'd first written it. In hindsight local variables is much less painful so I'll just do that.

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -535,22 +535,51 @@ - (void)_didFinishLoadingDataForCustomContentProviderWithSuggestedFilename:(cons
+
+- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    if (customContentViewDidHandleKeyEvent(event, [self isFirstResponder], _customContentView))
```
Nit - given that this helper function uses multiple states that are already on `WKWebView` (i.e. first responder status and `_customContentView`), I would just make this a helper method like
```- (BOOL)_tryToHandleKeyEventInCustomContentView:(UIPressesEvent *)event;```
...and then just use that in these three places.

aprotyas replied:
Changed this from a static function to a helper method

pxlcoder commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -1356,9 +1359,19 @@ static UIDeviceOrientation toUIDeviceOrientation(DeviceOrientation orientation)
+    }();
+    auto *press = [&keyboardEvent = keyboardEventDown] {
+        UIPressInfo *pressInfo = [[UIApplication sharedApplication] _pressInfoForPhysicalKeyboardEvent:keyboardEvent];
+        UIPress *press = [[UIPress alloc] init];
```
`RetainPtr` – this is a memory leak

aprotyas replied:
Wrapping it with `adoptNS()`, thanks!

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6503,13 +6502,75 @@ - (void)_handleKeyUIEvent:(::UIEvent *)event
+    bool isHardwareKeyboardEvent = !!event._hidEvent;
+    // We only want to handle key event from the hardware keyboard when we are
+    // first responder and we are not interacting with editable content.
+    if ([self isFirstResponder] && isHardwareKeyboardEvent && (_inputPeripheral || !_page->editorState().isContentEditable)) {
```
Hm...it also seems a tiny bit weird to run this logic in the case of `-pressesCancelled`, but I suppose that matches shipping behavior since `-_handleKeyUIEvent:` can be called for cancelled key events, so it's probably best to leave it here.

aprotyas replied:
Yeah I'm not quite sure what it _means_ to exercise this logic on a `pressesCancelled` but since this is shipping behavior I didn't probe too much here

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -6503,13 +6502,75 @@ - (void)_handleKeyUIEvent:(::UIEvent *)event
+        }
+        [super pressesEnded:presses withEvent:event];
+        return;
+    }
```
It would be nice to share some of this logic with the other 3 `-presses*` methods as well — maybe with a helper method like `- (BOOL)_handlePressesEvent:(UIPressesEvent *)event;`.

aprotyas replied:
Yup, made a helper here too!

```
- (BOOL)_tryToHandlePressesEvent:(UIPressesEvent *)event;
```

whsieh commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -1356,9 +1359,19 @@ static UIDeviceOrientation toUIDeviceOrientation(DeviceOrientation orientation)
-        kUIKeyboardInputModifierFlagsChanged, m_capsLockOn);
-    [[UIApplication sharedApplication] handleKeyUIEvent:keyboardEvent];
+    auto [keyboardEventDown, keyboardEventUp] = [capsLockOn = m_capsLockOn] {
+        auto *keyboardEventDown = createUIPhysicalKeyboardEvent(@"capsLock", [NSString string], capsLockOn ? UIKeyModifierAlphaShift : 0, kUIKeyboardInputModifierFlagsChanged, IsKeyDown::Yes);
```
Nit - `[NSString string]` can just be `@""`

aprotyas replied:
Fixed

whsieh commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -1356,9 +1359,19 @@ static UIDeviceOrientation toUIDeviceOrientation(DeviceOrientation orientation)
-    [[UIApplication sharedApplication] handleKeyUIEvent:keyboardEvent];
+    auto [keyboardEventDown, keyboardEventUp] = [capsLockOn = m_capsLockOn] {
+        auto *keyboardEventDown = createUIPhysicalKeyboardEvent(@"capsLock", [NSString string], capsLockOn ? UIKeyModifierAlphaShift : 0, kUIKeyboardInputModifierFlagsChanged, IsKeyDown::Yes);
+        auto *keyboardEventUp = createUIPhysicalKeyboardEvent(@"capsLock", [NSString string], capsLockOn ? UIKeyModifierAlphaShift : 0, kUIKeyboardInputModifierFlagsChanged, IsKeyDown::No);
```
(Ditto.)

aprotyas replied:
Fixed

whsieh commented on Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
```diff
@@ -605,11 +606,13 @@ static void setModifierFlagsForUIPhysicalKeyboardEvent(UIPhysicalKeyboardEvent *
     auto* keyboardEvent = [getUIPhysicalKeyboardEventClass() _eventWithInput:uiEventInputString inputFlags:inputFlags];
     setModifierFlagsForUIPhysicalKeyboardEvent(keyboardEvent, modifierFlags);
-    auto hidEvent = createHIDKeyEvent(hidInputString, keyboardEvent.timestamp, isKeyDown);
+    auto hidEvent = createHIDKeyEvent(hidInputString, keyboardEvent.timestamp, isKeyDown == IsKeyDown::Yes ? true : false);
```
Nit - this should also just be `isKeyDown == IsKeyDown::Yes`.

aprotyas replied:
Fixed

aprotyas committed:
Adopt UIResponder API methods in lieu of handleKeyUIEvent to respond to press events
https://bugs.webkit.org/show_bug.cgi?id=260217
rdar://100726289

Reviewed by Wenson Hsieh.

The handleKeyUIEvent: and _handleKeyUIEvent: SPI are both deprecated.
Instead, we choose to adopt the pressesBegan:, pressesChanged:,
pressesEnded:, and pressesCancelled: API in this commit. These methods
are forwarded events for handling in the same fashion as the SPI to be
removed.

* LayoutTests/fast/repaint/placeholder-after-caps-lock-hidden.html:

After the changes introduced in this patch, this test was dumping a lot
more painted rects. This was because previous key actions (`a`/`\b`)
were bleeding into the section of the test where we wanted to track
repaints (i.e. after the capslock indicator has been hidden).

We fix this by enforcing correct event ordering through appropriate async
usage of the UIHelper functions (i.e. awaiting `keyDown()` calls and the
resulting `keyUp` events).

* Source/WebKit/Platform/spi/ios/UIKitSPI.h:

1. Remove the deprecated SPI declarations.
2. Declare the UIPressInfo interface and some associated methods to
imbibe event information into a UIPress object.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _tryToHandleKeyEventInCustomContentView:]):

Helper function to determine whether an event needs forwarding based on
whether or not a custom content view handles it. Helps deduplicate code.

(-[WKWebView pressesBegan:withEvent:]):
(-[WKWebView pressesChanged:withEvent:]):
(-[WKWebView pressesEnded:withEvent:]):
(-[WKWebView pressesCancelled:withEvent:]):
(-[WKWebView _handleKeyUIEvent:]): Deleted.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _tryToHandlePressesEvent:]):

Helper function to determine whether an event needs forwarding based on
whether or not an input peripheral handles it. Helps deduplicate code.

(-[WKContentView pressesBegan:withEvent:]):
(-[WKContentView pressesEnded:withEvent:]):
(-[WKContentView pressesChanged:withEvent:]):
(-[WKContentView pressesCancelled:withEvent:]):
(-[WKContentView _handleKeyUIEvent:]): Deleted.

(WTR::UIScriptControllerIOS::toggleCapsLock):
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::createUIPhysicalKeyboardEvent):
(WTR::UIScriptControllerIOS::toggleCapsLock):

The `toggleCapsLock` implementation required two broad changes to
maintain the existing toggle behavior:

1. Simply calling `pressesBegan:` on the `UIApplication` singleton does
not suffice in getting our synthetic keyboard event sent to the content
view's overridden `pressesBegan:` definition. To work around this, we
manually create a `UIPress` object with knowledge of our synthetic
keyboard event, and then send the press/event combination directly to
our content view's `pressesBegan:` overload. This maintains the existing
toggle behavior.

2. In inspecting the press/event combinations received when toggling the
capslock key with a hardware keyboard, it was noticed that on iOS we send
both a keyDown and a keyUp event with the appropriate modifier flag
information whenever we want to toggle capslock, with the former being
sent to `pressesBegan:` and the latter being sent to `pressesEnded:`.
As such, we generate a pair of keyboard events and call the appropriate
`presses[Began/Ended]:` method directly on our content view to reflect
this behavior.

Canonical link: https://commits.webkit.org/267119@main
